gdctools/gdc_dice.py

```diff
@@ -164,17 +166,23 @@ def dice(self):
                     for tcga_id in tcga_lookup:
                         # Dice single sample files first
                         for file_d in viewvalues(tcga_lookup[tcga_id]):
-                            dice_one(file_d, trans_dict, raw_project_root,
+                            dice_one_status = dice_one(file_d, trans_dict, raw_project_root,
                                      diced_project_root, mfw,
                                      dry_run=self.options.dry_run,
                                      force=self.force)
+                            prog_status_tally[dice_one_status] += 1
 
                     #Then dice the multi_sample_files
                     for file_d in multi_sample_files:
-                        dice_one(file_d, trans_dict, raw_project_root,
+                        dice_one_status = dice_one(file_d, trans_dict, raw_project_root,
                                  diced_project_root, mfw,
                                  dry_run=self.options.dry_run,
                                  force=self.force)
+                        prog_status_tally[dice_one_status] += 1
+
+                # move completed metadata file to final location
+                # note it may include diced files that had errors
+                os.rename(diced_meta_file_partial, diced_meta_file)
```

Comment:
This code (and the comment directly above) should not execute when diced files had errors; by definition, the file should rename .partial;  this is grounds for not passing review.

Seems like all that's needed is:
```python
if 'error' not in prog_status_tally:
    os.rename(...)
```

Reply:
With the code as-is, a failure will cause attempts at running downstream steps to also fail.  With your suggested change, the downstream step will pass but silently use the wrong (old) data.

Reply:
Thanks for explaining, but I don't understand the logic/idea:  a .partial is supposed to indicate a failure state, and the absence of .partial represents a success state.  But you seem to be saying the opposite.  Downstream steps that observe upstream .partial (failure state) should fail immediately, no?

Reply:
Downstream steps look only for the main output file, not .partial, .error, etc. This respects encapsulation (Also, I was getting dizzy when considering how to correctly and intuitively handle all the permutations of various files existing vs not.) A .partial is currently being used just to ensure that downstream steps never get a file that has been control-C'ed in the middle of.  If some files are listed that do not exist, the next step will also fail with a nonzero exit code.

Reply:
So, in the case cited here, the metadata.partial is just atomicity for file-level operations, so we know if catastrophic event occurred (like ran out of disk space, or CTRL-C)?

Not quite sure what you're getting at about how not looking for .partial "respects encapsulation"?

But earlier you said by  removing .partial ONLY when all files have diced causes downstream steps to "pass" by silently using wrong (old) data?  But, if the only data that actually passed through the dicer is "old" then isn't that the correct thing to do?

Can you explain what a downstream failure looks like, then, if we keep the code "as is"? Does the gdc_loadfile (a) refuse to launch at all OR (b) launch and process everything it can, but skip the files that failed to dice (and flag them with warning messages, etc)?

Reply:
I did not edit gdc_loadfile as part of this, and it looks like it does not check for file existence.  So, the loadfile will be created fine; the failure message will arise when you go to upload missing files.

The issue that we were trying to address is someone running step a, ignoring a failure message (easy in bash), running step b (inadvertently on old data), seeing it pass, and charging ahead with analysis on a different set of data than they thought.

Reply:
Actually, allowing the loadfile to be created when the underlying data is not there makes the entire situation worse: because it will not be caught when gsutil is used to push data into cloud (because it acts like rsync, and doesn't what should or shouldn't be there), BUT rather will cause the merge pipelines to mysteriously fail (and this is vastly harder situation to debug).

Each of the tools output what version of data they're operating upon, e.g. from dicer log

```
2018-01-19 01:08:52,229[INFO]: Mirror date: 2018_01_19
```
 so I don't see the problem with charging ahead on old data--one is clearly notified-- and if one chooses to not pay attention to that notification then ... well, one is not being a careful analyst.

So, isn't the suggestion of using the .error or .partial on all files (data and metadata/json) a remedy for your concern?  If gdc_loadfile sees gdc_dice.error then it complains and stops.  If gdc_dice sees gdc_mirror.error it complains and stops.  etc.   What am I missing here?    You want to avoid users making inadvertent mistakes, and this is a way to help that...right?

Reply:
This failing case would be better if gdc_loadfile checked for the existence of every file before moving forward, but that would make the common (passing) case much slower.

In the end it is a matter of tradeoffs.  Streamline and bulletproof the use cases you care about most, and for the less central use cases try not to unpleasantly surprise the user.

Reply:
We may be talking past one another, or there are (unstated?) assumptions lurking in our perspectives.  In my view, if--when an error occurs--**gdc_dice** were to output **metadata.json.error** instead of **metadata.json** then downstream tools would not have to do anything at all, they just exit immediately.   You seem to be opposed to this, and suggest that it's "better" to check every file, but I don't understand why?  Because if there is no .error suffix then **gdc_loadfile** should not have to check any files, right?  The absence of .error from **gdc_dice** is the contract to **gdc_loadfile** that everything within the metadata.json indeed exists ... as **gdc_dice** will have already iterated over the entire suite of files (by necessity).

Reply:
1) With a .error file, we would need to handle the case where it coexists with a (no special extension) pass file and/or a .partial file... we don't have full control of the file system, and this case will arise somewhere, eg due to sloppy snapshot restores, if not from some race condition bug in our own code.  We want to make sure that at least we don't do the wrong thing, and we would want to have regression tests around it or the feature will surely regress.  Seems hard to make reliable.

2) With a .error file, or with a variant where the error condition is flagged by an extra row or column in the (no special extension) file, what action is the user supposed to do with it? We aren't expecting the user to go in and look at or (shudder) modify the metadata file.  There are no proposals to be able to -force the error file to be applied.  Once the issue is fixed, a .error or a error-marked (no special extension) file is still worthless, only waiting around to be redone.

3) The merger should reliably gripe if files are missing, either at the localization stage (if on Firecloud) or runtime (if running locally).  The remedy (using the code in this PR) would be to fix the dicing somehow, upload the missing files, and move forward with the same loadfiles.

If you address issue 1, making the system logic sufficiently reliable for your satisfaction, I have no real problem with using .error files.

Reply:
Thank you, Gordon.   Using 'date' as synonymous with 'tag' ...

1)  If a **date1**.error file of any kind exists then downstream tools which would consume that respective file cannot use that date.  Period. The tool will object to loading such and abort, so the user must either (a) clear that error (wherever it may occur upstream) or (b) try an earlier date.  In either case, user has full knowledge that a problem exists and full control over the remedy.
2)  See answer to (1)
3)  Loadfiles should not be generated that cannot work.  Period.  This is a straightforward example of the principle that it saves time & effort to ascertain errors as far upstream as possible, and fail, rather than generate unusable results and punt them downstream where the problem is even harder to debug (e.g. gdc_dice or gdc_loadfile, versus FireCloud merger).  Of course, a merger should fail if a file doesn't exist; that should be an almost impossible situation (e.g. where gsutil lies that a file has been uploaded, when it has't), but we should guard against it in the merger code.  What we should NOT do, though, is to casually allow such a condition to arise by generating unusable load files; that situation can be easily identified--during GDCtools processing, where such problems would be catalyzed--and is more effectively diagnosed and remedied there instead of downstream where there are more layers of tech stack to complicate debugging.